clouddrift/wavelet.py

```diff
@@ -0,0 +1,518 @@
+    """
+    m = 10000
+    omhigh = np.linspace(0, np.pi, m)
+    f = np.zeros_like(ga, dtype="float")
+
+    for i in range(0, len(ga)):
+        fm, _, _ = morsefreq(ga[i], be[i])
+        om = fm * np.pi / omhigh
+        lnwave1 = be[i] / ga[i] * np.log(np.exp(1) * ga[i] / be[i])
+        lnwave2 = be[i] * np.log(om) - om ** ga[i]
+        lnwave = lnwave1 + lnwave2
+        index = np.nonzero(np.log(eta) - lnwave < 0)[0][0]
+        f[i] = omhigh[index]
+
+    return f
+
+
+def morseprops(
+    ga: Union[np.ndarray, float],
+    be: Union[np.ndarray, float],
+) -> Union[Tuple[np.ndarray], Tuple[float]]:
+    """
+    Properties of the demodulated generalized Morse wavelets.
+
+    Parameters
+    ----------
+    ga: np.ndarray or float
+       Gamma parameter of the wavelet.
+    be: np.ndarray or float
+       Beta parameter of the wavelet.
+
+    Returns
+    -------
+    p: np.ndarray or float
+    skew: np.ndarray or float
+    kurt: np.ndarray or float
+    """
+    # test common size? or could be broadcasted
+    p = np.sqrt(ga * be)
+    skew = ga - 3 / p
+    kurt = 3 - skew**2 - 2 / p**2
+
+    return p, skew, kurt
+
+
+def morseafun(
+    ga: Union[np.ndarray, float],
+    be: Union[np.ndarray, float],
+    order: Optional[np.int64] = 1,
+    norm: Optional[str] = "bandpass",
+) -> float:
+    # add test for type and shape in case of ndarray
+    if norm == "energy":
+        r = (2 * be + 1) / ga
+        a = (
+            2 * np.pi * ga * (2**r) * np.exp(_lgamma(order) - _lgamma(order + r - 1))
+        ) ** 0.5
+    elif norm == "bandpass":
+        fm, _, _ = morsefreq(ga, be)
+        a = np.where(be == 0, 2, 2 / (np.exp(be * np.log(fm) - fm**ga)))
+    else:
+        raise ValueError(
+            "Normalization option (norm) must be one of 'energy' or 'bandpass'."
+        )
+
+    return a
+
+
+# def _gamma(
+#    x: Union[np.ndarray, float],
+# ) -> np.ndarray:
+#    """
+#    Returns gamma function values. Wrapper for math.gamma which is
+#    needed for array inputs.
+#    """
+#    # add test for type and shape in case of ndarray?
+#    if type(x) is np.ndarray:
+#        x_ = x.flatten()
+#        y = np.zeros_like(x_, dtype="float")
+#        for i in range(0, np.size(x)):
+#            y[i] = math.gamma(x_[i])
+#        y = np.reshape(y, np.shape(x))
+#    else:
+#        y = math.gamma(x)
+#
+#    return y
+
+
+# this maybe not useful
+# def _lgamma(
+#    x: Union[np.ndarray, float],
+# ) -> np.ndarray:
+#    """
+#    Returns logarithm of gamma function values. Wrapper for math.lgamma which is
+#    needed for array inputs.
+#    """
+#    # add test for type and shape in case of ndarray?
+#    if type(x) is np.ndarray:
+#        x_ = x.flatten()
+#        y = np.zeros_like(x_, dtype="float")
+#        for i in range(0, np.size(x)):
+#            y[i] = math.lgamma(x_[i])
+#        y = np.reshape(y, np.shape(x))
+#    else:
+#        y = math.lgamma(x)
+#
+#    return y
+
```

Comment:
If we're happy with SciPy we can remove this.

```suggestion
```

Reply:
done